mapstructure.go

```diff
@@ -192,6 +192,9 @@ func (d *Decoder) Decode(raw interface{}) error {
 func (d *Decoder) decode(name string, data interface{}, val reflect.Value) error {
 	if data == nil {
 		// If the data is nil, then we don't set anything.
+		if d.config.ZeroFields {
+			val.Set(reflect.Zero(val.Type()))
+		}
```

Comment:
Since we do call `val.Set()` here, I think this should also be added to `d.config.Metadata.Keys`.

```
if d.config.Metadata != nil && name != "" {
	d.config.Metadata.Keys = append(d.config.Metadata.Keys, name)
}
```